feat: Matrix Integration for EmbeddedChat

Aryan-Verma-999 opened the pull request:
# feat: Matrix Integration for EmbeddedChat
## Acceptance Criteria fulfillment
- [x] Integrate EmbeddedChat with the new Rocket.Chat native Matrix homeserver while keeping direct access to Rocket.Chat as an alternate option.
- [x] Introduce a specialized theme and layout setting called "WithMatrix Integration" to align the widget's style with modern Matrix clients (like Element).
- [x] Create a new theme file `MatrixTheme.js` (Primary Emerald Green #0dbd8b, Dark bg #15191e, Light bg #ffffff, and font family "Inter", "Helvetica Neue", sans-serif).
- [x] Add layoutMode prop supporting "bubble" | "timeline" where timeline mode removes message bubbles, groups consecutive messages under a single username/timestamp header, and highlights active messages on hover.
- [x]  Render a Globe Icon in the widget header instead of the standard # (hash) or lock icon for federated matrix channels.
- [x]  Render emerald-bordered server origin badges (e.g., [matrix.org]) next to federated users' names in the chat.
- [x]  Add Docusaurus documentation explaining how to configure the Matrix features and modes in EmbeddedChat.
Fixes #1029 
## Video/Screenshots
<img width="1919" height="986" alt="Screenshot_2026-07-12_20-02-11" src="https://github.com/user-attachments/assets/73784f4d-1590-48f8-ac71-0a9963cfd283" />

## PR Test Details
1.  Run `yarn storybook`.
2.  Go to the ** EmbeddedChat / WithMatrixt** story.
4.  Enter a valid  `host` url and rc `roomId`.
5.  Log in and verify message sending/receiving.

**Note**: The PR will be ready for live testing at https://rocketchat.github.io/EmbeddedChat/pulls/pr-1030 after approval. Contributors are requested to replace `<pr_number>` with the actual PR number.

Aryan-Verma-999 committed:
feat: Matrix Integration for EmbeddedChat (RC native federation)

Implements Matrix mode via RC native federation REST API.
Removes matrix-js-sdk dependency per architecture direction.

- mode prop switches between rocketchat and matrix modes
- RC handles Matrix federation transparently
- Full backward compatibility with rocketchat mode preserved

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/RocketChat/EmbeddedChat?pullRequest=1030) <br/>All committers have signed the CLA.

Aryan-Verma-999 commented:
Hi @Spiral-Memory , when you get time, could you take a quick look at this PR?

I've implemented the first pass for matrix integration

Pratheek555 commented:
Hey @Aryan-Verma-999!

If I am not wrong, there are some linting issues with the file **_packages/api/src/EmbeddedChatApi.ts_**,  please do give it a check. If I am mistaken, ignore this message.

Thank you!

Spiral-Memory commented:
Hi @Aryan-Verma-999,  
Could you please format it using Prettier? Thanks!

Aryan-Verma-999 commented:
Hi @Spiral-Memory !,
Thanks for the review! I've applied prettier formatting and verified that all linting checks pass. The changes have been pushed.
Please let me know if you need any other changes!

Spiral-Memory commented:
@Aryan-Verma-999 Could you please attach a demo video showcasing all the features you’ve implemented? This would help us understand it better. Great work, by the way!

Aryan-Verma-999 commented:
Hi @Spiral-Memory ,

I've recorded a demo video showcasing all the implemented features. Here's what's demonstrated:
##  Features Demonstrated
1. **Authentication & Connection**
   - Single-click login with Matrix credentials
   - Automatic connection to Matrix homeserver
   - Successful sync with server
2. **Room Management**
   - Auto-joining configured Matrix room
   - Room info display (name, avatar)
3. **Message History**
   - Loading existing messages on connect
   - Correct message ordering (newest at bottom)
   - Display names shown for all users
4. **Real-time Messaging**
   - Sending messages to Matrix room
   - Receiving messages in real-time
   - No duplicate messages
5. **Mode Switching**
   - Easy toggle between `rocketchat` and `matrix` modes via the `mode` prop


https://github.com/user-attachments/assets/d0d57123-620c-4134-9cd7-af022467dc0c



All core functionality is working as expected for public, unencrypted Matrix rooms.
**Note:** Encrypted rooms show a placeholder message ("⚠️ [Encrypted Message]") as E2EE support is not yet implemented.

Please let me know if you'd like me to demonstrate any specific scenarios or if you have any questions!

Spiral-Memory commented:
Thanks, @Aryan-Verma-999. I'll go through it and post the updates. Thanks a lot for your contribution.

Spiral-Memory commented:
The video looks great. In the description, you mentioned it can display avatar images and room avatars, so please share a screenshot of this with some pictures. It would also be helpful if the video could demonstrate two-way communication for better understanding, with both users being different, to see if it renders correctly all other modes. One small suggestion: instead of having a separate folder for matrix mode, consider naming it like the others, such as WithMatrix.

Great job so far; this might take some time for review and a few iterations, so be sure to collaborate.

Spiral-Memory commented:
I think you’re using the Matrix JS-SDK but we don’t want to log in directly with Matrix credentials. Instead, we want to use the Rocket.Chat-built homeserver. The current implementation might not correct, so please check the homeserver setup. I should be able to log in with Rocket.Chat credentials and still send messages to Matrix, with the homeserver possibly running on your system to make this work.

Please have a look at : https://github.com/RocketChat/homeserver

Aryan-Verma-999 commented:
Thanks for pointing this out. I’ll switch the implementation to use the Rocket.Chat homeserver instead of direct Matrix credential login. I’ll go through the homeserver repo, set it up locally, and adjust the flow so users authenticate with their Rocket.Chat credentials while still being able to send messages to Matrix rooms. I’ll update the PR once I have this aligned

Spiral-Memory commented:
Hi @Aryan-Verma-999 
Any update on it ?

Aryan-Verma-999 commented:
Hi @Spiral-Memory , sorry for the delay 

i had university exams over the past few weeks, and during that time i also reassessed the approach for matrix integration. while working on it, i realized that with rocket.chat's shift towards native federation (which is still in alpha) and the homeserver, the intended integration flow isnt clear yet, especially around message send and receive endpoints

i also couldn't find any concrete documentation around the homeserver expected client side integration, which made it hard to validate assumptions. to clarify this, i tried reaching out to gazzo and diego as well to understand the intended setup, but im still waiting to hear back

before pushing further changes i wanted to align on the expected direction. ive been looking into the homeserver side to understand how embeddedchat is expected to interact  with it at this stage 

im actively back on this now and would appreciate guidance on the preferred approach so i can continue in the right direction 

thanks for your patience 
sorry again for the delay

Spiral-Memory commented:
No worries, take your time

Spiral-Memory reviewed (changes requested):
Requesting changes as discussed

Aryan-Verma-999 commented:
Hi Zeeshan,

Quick update:

I've completed the implementation approach for integrating EmbeddedChat with Rocket.Chat's native federation.

However, during testing I discovered a server-side limitation in Rocket.Chat: when users self-join federated rooms (without being explicitly invited), they're not registered on the Matrix homeserver. This means their messages don't federate to remote Matrix users.

I've raised an issue in the Rocket.Chat repo explaining the problem and proposing solutions: [issue](https://github.com/RocketChat/Rocket.Chat/issues/38239)

Once that's resolved on the RC side, I'll push the EmbeddedChat changes. The client-side work is ready and waiting.

Let me know if you have any questions or want to discuss alternative approaches in the meantime.

Aryan-Verma-999 commented:
Hi @Spiral-Memory, quick update
The federation self-join bug I found on the RC side (RocketChat/Rocket.Chat#38239) is moving along. ggazzo picked it up and I'm working with him to help get it fixed.
Will circle back and update this PR once the upstream fix lands :)

Spiral-Memory commented:
> Hi @Spiral-Memory, quick update The federation self-join bug I found on the RC side ([RocketChat/Rocket.Chat#38239](https://github.com/RocketChat/Rocket.Chat/issues/38239)) is moving along. ggazzo picked it up and I'm working with him to help get it fixed. Will circle back and update this PR once the upstream fix lands :)

Sure, thanks @Aryan-Verma-999 
Let me know once this is ready for review, i am marking it as draft

Aryan-Verma-999 committed:
feat: remove matrix-js-sdk, replace with RC native federation

Rewrites Matrix mode to use RC native federation REST API.
Removes external matrix-js-sdk dependency entirely.

- Authenticate with RC credentials, not Matrix credentials
- Join federated rooms via RC REST API
- RC handles Matrix S2S federation transparently
- Trigger afterJoinRoom hook via REST join on connect

Closes #1029

Aryan-Verma-999 committed:
chore: cleanup and simplify Matrix integration

- Simplify WithMatrix story
- Strip stray comments and console.log
- Fix localhost URLs in stories
- Sync packages

Aryan-Verma-999 commented:
pushed a major update to this pr 
implementation has been completely rewritten to use RC's native federation, removing the matrix-js-sdk dependency entirely. users now authenticate with RC creds and RC handles the matrix federation layer transparently

keeping as draft until upstream self-join fix lands

Spiral-Memory commented:
Please resolve the conflicts

Aryan-Verma-999 committed:
merge: resolve conflicts with upstream/develop

- EmbeddedChatApi.getRoomInfo: delegate to channelInfo() (upstream)
- EmbeddedChatApi.getUsersInRole: keep {users:[]} catch fallback (Matrix)
- EmbeddedChatApi.getUserRoles: add new REST method from upstream
- EmbeddedChatApi.sendAttachment: auth guard + 2-step media upload (upstream)

Aryan-Verma-999 commented:
Conflicts resolved and rebased on latest develop.
While testing I also noticed image attachment previews returning 403 now, likely because of recent upload api change.
would you like me to open an issue + PR for this or wait until you clear the current backlog

Pratheek555 commented:
> Conflicts resolved and rebased on latest develop. While testing I also noticed image attachment previews returning 403 now, likely because of recent upload api change. would you like me to open an issue + PR for this or wait until you clear the current backlog

Yep noticed the same issue with the attachment previews.

Spiral-Memory commented:
> Conflicts resolved and rebased on latest develop. While testing I also noticed image attachment previews returning 403 now, likely because of recent upload api change. would you like me to open an issue + PR for this or wait until you clear the current backlog

Pls raise that and tag me on that mentioning "critical fix"

deepak0x commented:
Hey @Aryan-Verma-999 and @Spiral-Memory

While looking at the upstream self-join bug ([RocketChat/Rocket.Chat#38239](https://github.com/RocketChat/Rocket.Chat/issues/38239)), I've been researching if there's a way we can work around it on the EmbeddedChat side — without waiting for the RC fix to land.

Found two approaches that could unblock this PR:

---

**Option 1 — Self-Invite Pattern (no extra config needed)**

The root cause of the bug is that `channels.join` sets `inviter = null`, which causes the federation hook to skip homeserver registration. But `channels.invite` always has an `inviter` (the caller), which satisfies the hook condition and triggers proper Matrix registration.

The trick: when a user opens a federated room in EmbeddedChat, instead of calling `channels.join`, we call `channels.invite` with the user's own `userId` in the invite list. This makes RC think the user was invited (by themselves), the federation hook fires, and they get registered on the Matrix homeserver — so their messages federate correctly.

Verified this by reading the RC source directly:
- `beforeAddUserToRoom` hook checks `if (!inviter) return` — this is the exact line that blocks self-joiners
- `channels.invite` passes the calling user as `inviter` → condition passes → homeserver registration happens

Implementation would be a small change in `EmbeddedChatApi.ts` — detect if the room is federated on mount, and use the invite endpoint instead of join.

---

**Option 2 — Admin Bot Pre-Invite (opt-in, more reliable)**

EmbeddedChat could accept an optional `adminToken` prop. When a user loads a federated room, EmbeddedChat uses the admin token to invite the user before they send any message. This is 100% reliable and matches how most production federated rooms would be managed anyway.

---

Option 1 is the cleanest since it requires zero extra setup from the developer embedding the chat. Happy to implement either if it sounds good to you both — would love your thoughts on whether Option 1 is a safe assumption about RC's invite flow behavior.

Aryan-Verma-999 commented:
Hey @deepak0x , thanks for diving into this 
 
i actually tested option 1 earlier. the self-invite approach does pass the `!inviter` check, But there's another check in the same `beforeAddUserToRoom` federation hook that blocks it:
 
```
if (!isUserNativeFederated(user) && !(await FederationMatrix.canUserAccessFederation(user)))
  throw error-not-authorized-federation
```
`isUserNativeFederated` is always false for the users joining through self-invite so this check fails 
the self-invite trick only bypasses the inviter check but not this authorization gate, so option 1 won't work.

option 2 with adminToken would work since the admin already is natively federated and has access-federation, so the full hook runs correctly. 

i think the cleanest path is landing the server fix rather than adding client-side workarounds in EmbeddedChat

happy to pair on this if @Spiral-Memory  wants to move forward with option 2

deepak0x commented:
Sure would like to hear @Spiral-Memory thoughts too ;)
Also if you like i can help too in this. Have gone through the doc of rocket chat federation and got good understanding of it

deepak0x commented:
Quick correction on Option 1 — it fails earlier than the federation hook. The actual blocker is in `addUsersToRoomMethod`: regular users don't have `add-user-to-joined-room` or `add-user-to-any-c-room` permission, so it throws `error-not-allowed` before federation logic is ever reached. Option 1 is out.

**On Option 2 — I traced through the full code path and it works.**

When an admin or bot account calls `channels.invite`, the `inviter` field is not null, which is the key condition the federation hook requires. `canUserAccessFederation` passes for regular users since the `user` role has `access-federation` by default. `FederationMatrix.inviteUsersToRoom()` fires, the user gets registered on the Matrix homeserver, and their messages federate correctly.

The only concern is security — passing an admin token as a browser-side prop exposes it in DevTools. The safe approach is a dedicated bot account with only `add-user-to-any-c-room` permission assigned manually.

Implementation would be a new optional `adminToken` prop in EmbeddedChat. On mount, if the room is federated, EmbeddedChat calls `channels.invite` using that token before the user sends their first message. Happy to implement this if it sounds good.

Aryan-Verma-999 committed:
feat: transparent native matrix federation via joinRoom hook in EmbeddedChatApi

Aryan-Verma-999 committed:
feat: Matrix integration theme, timeline layout, and badges

Aryan-Verma-999 committed:
formatted with prettier

Spiral-Memory commented:
> Quick correction on Option 1 — it fails earlier than the federation hook. The actual blocker is in `addUsersToRoomMethod`: regular users don't have `add-user-to-joined-room` or `add-user-to-any-c-room` permission, so it throws `error-not-allowed` before federation logic is ever reached. Option 1 is out.
> 
> **On Option 2 — I traced through the full code path and it works.**
> 
> When an admin or bot account calls `channels.invite`, the `inviter` field is not null, which is the key condition the federation hook requires. `canUserAccessFederation` passes for regular users since the `user` role has `access-federation` by default. `FederationMatrix.inviteUsersToRoom()` fires, the user gets registered on the Matrix homeserver, and their messages federate correctly.
> 
> The only concern is security — passing an admin token as a browser-side prop exposes it in DevTools. The safe approach is a dedicated bot account with only `add-user-to-any-c-room` permission assigned manually.
> 
> Implementation would be a new optional `adminToken` prop in EmbeddedChat. On mount, if the room is federated, EmbeddedChat calls `channels.invite` using that token before the user sends their first message. Happy to implement this if it sounds good.

If not fixed by the core team, we may go with 2nd approach but using the EC app we have for secure admin token store

Aryan-Verma-999 commented:
https://github.com/user-attachments/assets/e36fe572-0be3-4d69-a2d3-2cee2ebc704e

demo-matrix-federation

Spiral-Memory reviewed (approved):
This is a good start, but simply adding changes to the existing theme pattern and minor badges isn’t enough, so more work is needed. I’m merging this for now, but I’m expecting a solid matrix-specific design and the business features we discussed. Thanks.

Spiral-Memory commented:
<img width="664" height="359" alt="image" src="https://github.com/user-attachments/assets/d1c1838a-2fc4-4543-b9d6-291078e1c09f" />


Can you add the reference to this new doc here in same format in root README.md

Aryan-Verma-999 commented:
Thanks for merging, Sure I'll follow up with dedicated PRs for both solid-matrix specific design and business features.
will scope them out and implement them

Aryan-Verma-999 commented:
> <img alt="image" width="664" height="359" src="https://private-user-images.githubusercontent.com/78961432/620564071-d1c1838a-2fc4-4543-b9d6-291078e1c09f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODM4NzMyMzMsIm5iZiI6MTc4Mzg3MjkzMywicGF0aCI6Ii83ODk2MTQzMi82MjA1NjQwNzEtZDFjMTgzOGEtMmZjNC00NTQzLWI5ZDYtMjkxMDc4ZTFjMDlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA3MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNzEyVDE2MTUzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNhZTY4YWU0OGU1ZDI2NmIyODk0ODI4NTViNTQyMzkyZGQyYmExYTk1MGZhNDg0YjQ4ZjdkYzE0YzgyOTlkMjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.ky2_Zo_vkx1XeWggqZylyQV13NBtemY_kXVFcoA14UM">
> 
> Can you add the reference to this new doc here in same format in root README.md

sure

Aryan-Verma-999 committed:
doc: add Matrix Integration reference to main README

Spiral-Memory commented on packages/docs/docs/Usage/matrix.md:
```diff
@@ -0,0 +1,95 @@
+
+---
+
+## Architectural Details
```
Keep a different title, this is under Usage section, so explain from the perspective of user integrating the widget

Aryan-Verma-999 replied:
changed

Spiral-Memory commented on packages/docs/docs/Usage/matrix.md:
```diff
@@ -0,0 +1,95 @@
+- **Header Pill**: If a room contains federation metadata, a globe icon alongside a green pill (`matrix:origin-server`) is rendered in the `ChatHeader`.
+- **Message Sender Badge**: External federated users are identified using the format `@username:homeserver.org`. If the sender's homeserver differs from the local homeserver, a subtle green outlined badge displaying the remote domain (e.g. `matrix.org`) is shown next to their display name.
+
+### 4. Dynamic Federation Joining
```
This is not required to include

Aryan-Verma-999 replied:
sure, removing it

Spiral-Memory commented on packages/docs/docs/Usage/matrix.md:
```diff
@@ -0,0 +1,95 @@
+
+---
+
+## Testing in Development
```
Remove this as well

Aryan-Verma-999 replied:
done

Aryan-Verma-999 committed:
doc: update matrix.md based on review feedback
